lib/rspec/core/memoized_helpers.rb

```diff
@@ -123,14 +123,40 @@ def is_expected
 
       # @private
       def __memoized
-        @__memoized ||= {}
+        @__memoized ||= Memoized.new
+      end
+
+      # @private
+      class Memoized
+        def initialize
+          @memoized = {}
+          @mutex    = Mutex.new
+        end
+
+        def for(key, &initializer)
+          return @memoized[key][0] if @memoized.key? key
+
+          value = initializer.call
```

Comment:
Without mixing the two, this should be correct behaviour: It is true that a nondeterministic value can be memoized, but not true that a nondeterministic value can be returned. Ie **for two threads**, the first one to get to the synchronize block wins, so second thread will discard its calculated value and all threads see the same result.
**Within the same thread** the last one to the memoized block wins so that you can stack calls to let.

We have to invoke the block outside the mutex because if the block calls super and hits another let block, then it will end up back in this same method, basically this:

```ruby
m = Mutex.new
m.synchronize { m.synchronize { } }
```

Though as I think through it now, I can see two issues:

1. If you mix these two (call `super()` to another let block, with multiple threads), it could do the wrong thing, seeing the memoized value of the supercall from the other thread. Ideally you could lock `let` such that only one thread could calculate it, but you run into the issue described above. There's `Mutex#owned?` which might be useful, but it says "This API is experimental, and subject to change" :/
1. If the calls to let have side effects, you can incur those side effects multiple times, even though both threads will see the same value come back from the let block. Again, the ability to lock to a thread seems like it would fix this.

I have a sense that we can make it work in all these strange cases, and also be simpler. Gonna do right-brain things for a bit and see if it manifests.